src/backtrace/dbghelp.rs

```diff
@@ -90,12 +90,38 @@ struct MyContext(CONTEXT);
 
 #[inline(always)]
 pub unsafe fn trace(cb: &mut dyn FnMut(&super::Frame) -> bool) {
+    let thread = GetCurrentThread();
+    trace_thread(cb, thread)
+}
+
+#[inline(always)]
+pub unsafe fn trace_thread(cb: &mut dyn FnMut(&super::Frame) -> bool, thread: HANDLE) {
     // Allocate necessary structures for doing the stack walk
     let process = GetCurrentProcess();
-    let thread = GetCurrentThread();
 
     let mut context = mem::zeroed::<MyContext>();
-    RtlCaptureContext(&mut context.0);
+    if thread == GetCurrentThread() || thread.is_null() {
+        RtlCaptureContext(&mut context.0);
+    } else {
+        // The suspending and resuming of threads is very risky.
+        // It can end in a deadlock if the current thread tries to access
+        // an object thats been locked by the suspended thread.
+        // That's why we only do as little work as possible while
+        // the thread is suspended, and resume it quickly after.
+
+        // There is most definitely more pitfalls i haven't thought
+        // of or encountered. This is windows after all.
+
+        context.0.ContextFlags = CONTEXT_ALL; // TODO: Narrow down required flags
+        if SuspendThread(thread) as i32 == -1 {
+            ResumeThread(thread);
+            return;
+        }
+        let status = GetThreadContext(thread, &mut context.0);
+        if ResumeThread(thread) as i32 == -1 || status == 0 {
```

Comment:
What if the other thread was already suspended? I think it shouldn't resume it in that case.

Reply:
I think `SuspdendThread` in the above lines suspends the thread. We have to resume it here after we have extracted to context.

Reply:
Also see 
https://github.com/tikv/pprof-rs/issues/9#issuecomment-1114222675

Reply:
What if someone called `SuspendThread` and then tried to get a backtrace? The current code would resume the thread again, while it did be reasonable to expect that the thread stays suspended if you suspended it yourself.

Reply:
`ResumeThread` does not resume the thread, it increments the suspend count.
In this code, `SuspendThread` increments the suspend count in line 116, and `ResumeThread` decrements it in line 121 or 117. Each code path has a pair of suspend/resume.   

See this for more information
https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-resumethread

Reply:
I didn't know that.